pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -155,58 +199,96 @@ func (p *Propagator) processEvent() {
 				},
 				Namespace: accessor.GetNamespace(),
 			},
-			owners: accessor.GetOwnerReferences(),
+			dependents: make(map[*node]struct{}),
+			owners:     accessor.GetOwnerReferences(),
 		}
-		p.addToGraphLookup(newNode)
-		ownerExists := p.updateOwners(newNode)
+		p.insertNode(newNode)
+		ownerExists := p.addToOwners(newNode)
 		// Push the object to the dirty queue if none of its owners exists in the Graph.
+		// TODO: when handling such an object in the dirtyQueue, the worker should
+		// enqueue an Add event to the eventQueue if the owner exists. This is
+		// to prevent the following case: GC observes the creation of dependent
+		// D, but hasn't observed the creation of Owner O yet; GC adds D to the
+		// dirtyQueue, worker works on D, and O exists in the API server, so GC
+		// doesn't remove D. Then, Propagator sees the creation of O, but no one
+		// will update the graph to reflect that D is the dependent of O, so
+		// when O is deleted, D will not be cascadingly deleted.
 		if !ownerExists && len(newNode.owners) != 0 {
 			fmt.Println("CHAO: directly added to dirty queue: ", newNode.identity)
 			p.gc.dirtyQueue.Add(newNode)
 		}
 
-	case Update:
-		node, ok := p.graphLookup[accessor.GetUID()]
+	case updateEvent:
+		node, ok := p.uidToNode[accessor.GetUID()]
 		if !ok {
 			glog.Errorf("received an update for %v, but cannot find the node in the graph", accessor.GetUID())
```

Comment:
This really shouldn't happen in the current system, so I prefer explicitly record the error.